### DISCUSSION TITLE
Parsing/Coloring Ansible Status

### DISCUSSION THREAD

**Discussion by OnceUponALoop** [xavierog/moulti#17] [Q&A] [answered]
This is an awesome project, it scratches an itch i've failed to scratch :)

Would it be possible to parse the ansible states and color accordingly? I would like to differentiate between the various states at a glance, ex. skipped (blue/grey), changed (yellow/orange), failed (red), OK/Good (Green).

I looked at the ansible callback code but nothing jumped out at me, can you point me in the right direction?

**Comment by xavierog** [comment]
Hi!

> Would it be possible to parse the ansible states and color accordingly? I would like to differentiate between the various states at a glance, ex. skipped (blue/grey), changed (yellow/orange), failed (red), OK/Good (Green).

Could you be more specific as to what you want to change? Currently, Ansible+Moulti looks like this, i.e. both Ansible lines and Moulti steps get colored based on status (skipped/changed/failed/success):
![moulti-ansible-colors-0](https://github.com/user-attachments/assets/1c7fe30c-ea42-438f-9116-50f3a7af8108)

That being said, there is room for improvement regarding diff (`-D`) and maybe JSON syntax highlighting.

> I looked at the ansible callback code but nothing jumped out at me, can you point me in the right direction?

Moulti implements an "stdout callback plugin". Such a plugin can indeed format and reformat the events emitted by Ansible. However, there is not much to see in `src/moulti/ansible/moulti.py` simply because our `CallbackModule` keeps the formatting made by `DefaultCallbackModule` and focuses on altering the `Display` class so that Ansible's output gets piped to Moulti instead of being printed to stdout.

**Reply by OnceUponALoop** [reply]
Man you're on top of things, I didn't expect a response for a while, let alone a few hours later :clap: 

> Could you be more specific as to what you want to change?

Sorry I should have been clearer, the question was more generic about color changes and so on but concretely I would like to add a blue/grey/other color for the "skipped" tasks, how hard would that be?

**Reply by xavierog** [reply]
> Sorry I should have been clearer, the question was more generic about color changes and so on but concretely I would like to add a blue/grey/other color for the "skipped" tasks, how hard would that be?

Although Moulti lets you [define your own classes/colors](https://github.com/xavierog/moulti/blob/3e10843eb4b83fb7fd9ffaf284c79330c9c38470/Documentation.md#how-to-define-my-own-step-classes-), what you ask definitely requires to patch Moulti.
Specifically, one needs to:

1. [introduce a class named "inactive"](https://github.com/xavierog/moulti/commit/8587e3746279053972823db1809478a943268771)
2. [revamp the class selection code in Moulti's Ansible callback plugin](https://github.com/xavierog/moulti/commit/c7d84750752bdcd6c5def6e11aa3da2f8bbd8d67)

All of this is available in the `devel` branch, which you can switch to through `pipx install git+https://github.com/xavierog/moulti.git@devel`. The contents of that branch should be released next Sunday.

**Reply by xavierog** [reply, marked as answer]
[Moulti 1.20.0 was released](https://github.com/xavierog/moulti/releases/tag/v1.20.0).

**Reply by OnceUponALoop** [reply]
I'm so sorry about not responding, I was under the assumption that I had already did!

I couldn't have imagined a response and modification this fast, thank you :fireworks: 

The skipped entries are now easily distinguishable from the rest - you're awesome :+1:
